### DISCUSSION TITLE
Questions around self-hosting

### DISCUSSION THREAD

**Discussion by sebastian** [ubicloud/ubicloud#1047] [General]
Hi

I discovered Ubicloud while searching for managed Postgres providers. Ubicloud is quite exciting and I see lots of potential for putting it to use.

I have created an account in your hosted service and have also been running the local demo version too. It appears they are not running the same version of the software, at the very least, the Postgres DB is not an option in the Ubicloud instance I see running locally. Are the versions different, or is it just selectively showing different services based on what hardware is available?

Presumably, your main focus is to offer the managed and hosted offering, as there you can make some money and run a sustainable business. I do hope at some point in the future there will be a way of more easily configuring Ubicloud for personal use or use within an organization.

For example, I would love to be able to:
- disable self-registration
- alter the types of VMs that can be created (for example I would love to be able to aggressively oversubscribe a host and offer shared and fractional CPU cores, be able to offer low storage VMs, etc)
- set another default username
- etc etc

I did see the CPU sizes being set in https://github.com/ubicloud/ubicloud/blob/main/lib/option.rb#L35-L37, but altering the values did not affect what I was allowed to create in the interface. Also, altering the source code has me concerned that the upgrade path is going to get more complicated down the line as I want to git-pull and get a later version of the system.

Thanks!

**Comment by byucesoy** [comment]
Hi @sebastian, thanks for your interest at Ubicloud and managed Postgres service. I'm the primary author of the Postgres service. I assume you are using /demo/docker-compose.yml for your local environment. Indeed, we didn't update it with Postgres service yet. The primary reason is that Postgres service itself needs other services to be set up first, namely a DNS service (for user friendly connection strings and failovers) and a blob storage (for backups).

Source code for these services along with Postgres service is available in our repository already, but it requires some set up and configuration for glueing everything together, which we didn't do in the demo yet. We are planning to add it soon, but there are few other things in the Postgres service we want to release first.

**Reply by sebastian** [reply]
Looking forward to seeing how this develops! Thank you for your answer and the details!

**Comment by ozgune** [comment]
Hey @sebastian - thank you for getting in touch with us!

I'm inlining my replies to your other questions below. We'll also share thoughts / questions in this discussion after the holidays.

> Presumably, your main focus is to offer the managed and hosted offering, as there you can make some money and run a sustainable business. I do hope at some point in the future there will be a way of more easily configuring Ubicloud for personal use or use within an organization.

Yes, our focus for the next 6-12 months will our managed offering. The primary reason is that the customer feedback cycle on our managed offering is faster. As we release more cloud services and they mature, our future goal will include making Ubicloud easier to use as a personal project or as part of an organization.

> For example, I would love to be able to: ... 

Thank you for the product feedback. If you'd like these feature requests to be standalone, and they are important to you, please feel free to open a GitHub issue. We'll then keep these in our backlog and evaluate their priority in upcoming product planning cycles.

> I did see the CPU sizes being set in https://github.com/ubicloud/ubicloud/blob/main/lib/option.rb#L35-L37, but altering the values did not affect what I was allowed to create in the interface. Also, altering the source code has me concerned that the upgrade path is going to get more complicated down the line as I want to git-pull and get a later version of the system.

Adding @fdr, @byucesoy, @enescakir in case they know why this didn't work. I'll let one of them chime in.

Ack on your point about forking the project making the upgrade path more complicated. Hopefully, Ubicloud will evolve to be either extensible enough for new features (this is hard), we'll start merging features from users, or implement the more important feature requests coming in from our users/customers.

**Reply by sebastian** [reply]
> The primary reason is that the customer feedback cycle on our managed offering is faster

Makes a lot of sense. And it's so much easier to fix problems and learn what works and what does not on a system you control yourself as opposed to one running on-prem somewhere else (we used to sell on-prem software to enterprises and debugging what a nightmare).

> Thank you for the product feedback. If you'd like these feature requests to be standalone, and they are important to you, please feel free to open a GitHub issue. We'll then keep these in our backlog and evaluate their priority in upcoming product planning cycles.

I'll happily add some issues – y ou can take it from there as it suits you and your roadmap. And of course, I fully understand that you aren't committing to anything :) 

👍🏻

**Comment by enescakir** [comment]
> > I did see the CPU sizes being set in https://github.com/ubicloud/ubicloud/blob/main/lib/option.rb#L35-L37, but altering the values did not affect what I was allowed to create in the interface. Also, altering the source code has me concerned that the upgrade path is going to get more complicated down the line as I want to git-pull and get a later version of the system.
> 
> Adding @fdr, @byucesoy, @enescakir in case they know why this didn't work. I'll let one of them chime in.

I assume you initiate the local version from the `./demo` directory. The docker-compose template located there [pulls the latest image from DockerHub](https://github.com/ubicloud/ubicloud/blob/6dafda77bc375346be698b40f1009bdc910c49a1/demo/docker-compose.yml#L29). It doesn't use the local code. If you wish to modify the code locally, you'll need to rebuild the Docker image. You should change the image section as follows in the `./demo/docker-compose.yml` file:

```yml
    image: ubicloud:dev
    build:
      context: ../
```

**Reply by sebastian** [reply]
Ah, of course. That's pretty obvious in hindsight. Thank you!

Yes indeed I was running the local version as provided in the demo folder (without looking all too careful at the docker-compose file.
